provider/resources.go

```diff
@@ -454,14 +456,82 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
 //go:embed cmd/pulumi-resource-gcp/bridge-metadata.json
 var metadata []byte
 
+// A predicate function that always returns true.
+func always[T any](T) bool { return true }
+
+// fixEmptyLabels applies a state edit to fix
+// https://github.com/pulumi/pulumi-gcp/issues/2372.
+func fixEmptyLabels(_ context.Context, req shimv2.PlanStateEditRequest) (cty.Value, error) {
+	// actualLabels holds the labels that we expect to be applied. It is the union of
+	// labels (from the resource) and defaultLabels (from the provider)
```

Comment:
This does not look correct - the resource might have unmanaged labels which we'd clobber if we apply this.

Reply:
This is more a naming issue then a semantic issue. `actualLabels` holds the labels that we expect to be applied. It is not the set of labels that we will apply. It is used only to manage unknowns.